deepmd/dpmodel/fitting/general_fitting.py

```diff
@@ -452,13 +447,14 @@ def _call_common(
             outs = self.nets[()](xx)
             if xx_zeros is not None:
                 outs -= self.nets[()](xx_zeros)
-        outs = xp.astype(outs, get_xp_precision(xp, "global"))
         outs += xp.reshape(
-            xp.take(self.bias_atom_e, xp.reshape(atype, [-1]), axis=0),
+            xp.take(
+                xp.astype(self.bias_atom_e, outs.dtype), xp.reshape(atype, [-1]), axis=0
+            ),
             [nf, nloc, net_dim_out],
         )
         # nf x nloc
         exclude_mask = self.emask.build_type_exclude_mask(atype)
         # nf x nloc x nod
         outs = xp.where(exclude_mask[:, :, None], outs, xp.zeros_like(outs))
-        return {self.var_name: xp.astype(outs, get_xp_precision(xp, "global"))}
+        return {self.var_name: xp.astype(outs, descriptor.dtype)}
```

Comment:
```suggestion
        return {self.var_name: outs)}
```
not cast should work